lib/packagemap.dart

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library package_config.packagemap;
+
+class Packages {
+  static const int _EQUALS = 0x3d;
+  static const int _CR = 0x0d;
+  static const int _NL = 0x0a;
+  static const int _NUMBER_SIGN = 0x23;
+
+  final Map<String, Uri> packageMapping;
+
+  Packages(this.packageMapping);
+
+  /// Resolves a URI to a non-package URI.
+  ///
+  /// If [uri] is a `package:` URI, the location is resolved wrt. the
+  /// [packageMapping].
+  /// Otherwise the original URI is returned.
+  Uri resolve(Uri uri) {
+    if (uri.scheme.toLowerCase() != "package") {
+      return uri;
+    }
+    if (uri.hasAuthority) {
+      throw new ArgumentError.value(uri, "uri", "Must not have authority");
+    }
+    if (uri.path.startsWith("/")) {
+      throw new ArgumentError.value(uri, "uri",
+      "Path must not start with '/'.");
+    }
+    // Normalizes the path by removing '.' and '..' segments.
+    uri = uri.normalizePath();
+    String path = uri.path;
+    var slashIndex = path.indexOf('/');
+    String packageName;
+    String rest;
+    if (slashIndex < 0) {
+      packageName = path;
+      rest = "";
+    } else {
+      packageName = path.substring(0, slashIndex);
+      rest = path.substring(slashIndex + 1);
+    }
+    Uri packageLocation = packageMapping[packageName];
+    if (packageLocation == null) {
+      throw new ArgumentError.value(uri, "uri",
+      "Unknown package name: $packageName");
+    }
+    return packageLocation.resolveUri(new Uri(path: rest));
+  }
+
+  /// Parses a `packages.cfg` file into a `Packages` object.
+  ///
+  /// The [baseLocation] is used as a base URI to resolve all relative
+  /// URI references against.
+  ///
+  /// The `Packages` object allows resolving package: URIs and writing
+  /// the mapping back to a file or string.
+  /// The [packageMapping] will contain a simple mapping from package name
+  /// to package location.
+  static Packages parse(String source, Uri baseLocation) {
```

Comment:
It can, I don't necessarily think it's better (or worse).
It currently follows the style used by other "parse" functions in being a static function on the type it returns.
If it becomes top-level, it needs a more telling name, like `parsePackages`, in which case the character-count saving is minuscule, and you lose the obviousness of what it returns.